### PR TITLE
Add APIs, CLIs, Unit Tests and functions for VM Schedule Snapshot

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
```06:07:12 File "xapi_xenops.ml", line 669, characters 5-24:
06:07:12 Error: This record expression is expected to have type
06:07:12          Xenops_interface.Vgpu.gvt_g
06:07:12        The field monitor_config_file does not belong to type Xenops_interface.Vgpu.gvt_g````

I think this is because of the new work that @johnelse has done on GVT-g that hasn't hit the yum repo yet for xcp-idl. I've restarted the travis job as it might be in now.

**Comment by simonjbeaumont** [comment]
@xen-git: retest this please.

**Comment by simonjbeaumont** [comment]
Confirmed this builds locally using xenserver-build-env using `-b trunk-ring3`.

**Comment by simonjbeaumont** [comment]
Wait, this is a pull-request for `feature-vmss` which is synced with `trunk-vmss` internally. This will fail to build if merged and it needs to be rebased in light of recent changes in xcp-idl.

I've just tested this using:
```
xenserver-build-env$ ./run.py -b trunk-vmss -p xapi ...
sh-4.2$ sudo su
[root@549de09885dc builder]# cd /external/code/xen-api
[root@549de09885dc xen-api]# configure && make
...
File "xapi_xenops.ml", line 669, characters 5-24:                                                                    
Error: This record expression is expected to have type
         Xenops_interface.Vgpu.gvt_g
       The field monitor_config_file does not belong to type Xenops_interface.Vgpu.gvt_g
```

**Comment by sharady** [comment]
I have done a b2b from trunk to trunk-vmss which should have all latest changes. Will build this PR once more.

**Comment by sharady** [comment]
@xen-git: retest this please.

**Comment by sharady** [comment]
@xen-git retest this please

**Comment by sharady** [comment]
@xen-git: retest this please.

**Comment by simonjbeaumont** [comment]
All CI passing now, removing the needs-updating flag. These PRs still need some level of review though.

File: ocaml/test/test_map_check.ml [hunk 1]
```diff
@@ -116,10 +116,114 @@ module Accessors = Generic.Make(struct
...
+let string_of_ty = function
+	| String -> "String"
+	| _ -> ""
+
+let string_of_ks ks =
+	let field,kss = ks in
+	List.map (fun (a, b) ->
+		let inner_string = List.map (fun (c, d) -> let e, f = d in
+		c ^ "," ^ (string_of_ty e) ^ f ) b |> String.concat ";" in
+		"[" ^ a ^ "," ^ "[" ^ inner_string ^ "]]"
```
Comment (lindig) [inline ocaml/test/test_map_check.ml]: Would using `Printf.sprintf` be simpler?

**Comment by simonjbeaumont** [comment]
These tests can only leave us in a better situation than we're currently in.

**Pull request merged**